{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Notebook for visualizing MVP phase 2 coherence check results\n",
    "\n",
    "Coherence check includes results generated by FastQC, Samtools Flagstat, and RTG Vcfstats."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import StringIO\n",
    "\n",
    "import pandas as pd\n",
    "\n",
    "import numpy as np\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "import google.datalab.bigquery as bq"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# MVP Phase 2 FastQC Results (n=1902)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mvp_fastqc_table=\"\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Per base quality"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "per_base_qual = bq.Query('SELECT index, ' +\n",
    "                         'ROUND(CAST(value AS FLOAT64), 1) AS value_tenth, ' +\n",
    "                         'COUNT(*) AS observations ' +\n",
    "                         'FROM `' + mvp_fastqc_table + '` ' +\n",
    "                         'WHERE dimension = \"per_base_qual\" ' +\n",
    "                         'GROUP BY index, value_tenth').execute().result().to_dataframe()\n",
    "per_base_qual.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "start_end = pd.DataFrame(per_base_qual['index'].str.split('-',1).tolist(), columns=['start', 'end'])\n",
    "per_base_qual = per_base_qual.join(start_end)\n",
    "\n",
    "per_base_qual[['value_tenth']] = per_base_qual[['value_tenth']].apply(pd.to_numeric)\n",
    "per_base_qual[['start']] = per_base_qual[['start']].apply(pd.to_numeric)\n",
    "per_base_qual[['observations']] = per_base_qual[['observations']].apply(pd.to_numeric)\n",
    "\n",
    "per_base_qual.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "per_base_qual[['log_observations']] = np.log(per_base_qual[['observations']])\n",
    "per_base_qual.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plot = per_base_qual.plot(kind='scatter', x='start', y='value_tenth', c='log_observations', s=35)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "per_base_qual.plot.hexbin(x='start', y='value_tenth', C='log_observations', reduce_C_function=np.max, gridsize=50)\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots()\n",
    "scat = per_base_qual.plot(kind='scatter', x='start', y='value_tenth', c='log_observations', cmap='plasma', ax=ax)\n",
    "ax.set_xlim(0,140)\n",
    "ax.set_title(\"Distribution of Per-base Sequence Quality across Genomes\")\n",
    "ax.set_xlabel(\"Base position\")\n",
    "ax.set_ylabel(\"Quality score\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## GC Content"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dimension = \"gc_content\"\n",
    "gc_content = bq.Query('SELECT sample, ' +\n",
    "                      'dimension, ' +\n",
    "                      'CAST(value AS INT64) as value ' +\n",
    "                      'FROM `' + mvp_fastqc_table + '` ' +\n",
    "                      'WHERE dimension = \"' + dimension + '\"').execute().result().to_dataframe()\n",
    "gc_content.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gc_content_counts = gc_content.groupby('value')['sample'].nunique()\n",
    "print gc_content_counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots()\n",
    "bar = gc_content_counts.plot.bar(alpha=0.5)\n",
    "ax.set_title(\"Distribution of Reads GC Content by Genome\")\n",
    "ax.set_xlabel(\"GC Content (%)\")\n",
    "ax.set_ylabel(\"Frequency\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Average sequence quality"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dimension = \"seq_quality\"\n",
    "avg_seq_qual = bq.Query('SELECT sample, ' +\n",
    "                        'SUM(CAST(index as INT64) * CAST(value AS FLOAT64))/SUM(CAST(value AS FLOAT64)) ' +\n",
    "                        'FROM `' + mvp_fastqc_table + '` ' +\n",
    "                        'WHERE dimension = \"' + dimension + '\" ' +\n",
    "                        'GROUP BY sample').execute().result().to_dataframe()\n",
    "avg_seq_qual.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.figure()\n",
    "plt.boxplot(avg_seq_qual['f0_'], 0, 'gD')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots()\n",
    "avg_seq_qual.hist(alpha=0.5, ax=ax)\n",
    "ax.set_title(\"Distribution of Average Read Qualities\")\n",
    "ax.set_xlabel(\"Average Read Quality\")\n",
    "ax.set_ylabel(\"Frequency\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Sequence length"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dimension = \"seq_len\"\n",
    "seq_len = bq.Query('SELECT sample, dimension, ' +\n",
    "                   'CAST(value AS INT64) as value ' + \n",
    "                   'FROM `' + mvp_fastqc_table + '` ' + \n",
    "                   'WHERE dimension = \"' + dimension + '\"').execute().result().to_dataframe()\n",
    "seq_len.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "seq_len_counts = seq_len.groupby('value')['sample'].nunique()\n",
    "print seq_len_counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots()\n",
    "bar = seq_len_counts.plot.bar(alpha=0.5)\n",
    "ax.set_title(\"Distribution of Read Lengths\")\n",
    "ax.set_xlabel(\"Read Length\")\n",
    "ax.set_ylabel(\"Frequency\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# MVP Phase 2 VCFStats Results (n=1902)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mvp_vcfstats_table = \"\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## SNPs per genome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "snps = bq.Query('SELECT sample, dimension, ' +\n",
    "                'CAST(value AS INT64) as value ' +\n",
    "                'FROM `' + mvp_vcfstats_table + '` ' +\n",
    "                'WHERE dimension = \"snps\"').execute().result().to_dataframe()\n",
    "snps.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots()\n",
    "hist = snps.hist(alpha=0.5, ax=ax)\n",
    "ax.set_title(\"Distribution of SNPs per Genome\")\n",
    "ax.set_xlabel(\"SNP Count\")\n",
    "ax.set_ylabel(\"Frequency\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Check number of variants in each category\n",
    "I am not sure whether RTG Tools vcfstats considers failed variants, and their documentation of vcfstats is very poor. Eye-check of ~10 vcfstats files does not show any variants marked as \"Failed Filters\".\n",
    "\n",
    "I'm going to run a query to see if any were marked as failing."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dimension_sums = bq.Query('SELECT dimension, SUM(CAST(value AS INT64)) AS sum ' +\n",
    "                'FROM `' + mvp_vcfstats_table + '` ' +\n",
    "                'WHERE dimension IN (\"failed_filters\", \"passed_filters\") ' +\n",
    "                'GROUP BY dimension').execute().result().to_dataframe()\n",
    "dimension_sums.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Indels per genome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dimension = \"indels\"\n",
    "indels = bq.Query('SELECT sample, dimension, ' +\n",
    "                  'CAST(value AS INT64) as value ' +\n",
    "                  'FROM `' + mvp_vcfstats_table + '` ' +\n",
    "                  'WHERE dimension = \"' + dimension + '\"').execute().result().to_dataframe()\n",
    "indels.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots()\n",
    "hist = indels.hist(alpha=0.5, ax=ax)\n",
    "ax.set_title(\"Distribution of INDELs per Genome\")\n",
    "ax.set_xlabel(\"INDEL Count\")\n",
    "ax.set_ylabel(\"Frequency\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "box = indels.plot.box()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.figure()\n",
    "plt.boxplot(indels['value'], 0, 'gD')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Cumulative insertions, deletions, and indels\n",
    "RTG Tools vcfstats outputs data on three categories of indels: insertions, deletions, and indels. From this google group post (https://groups.google.com/a/realtimegenomics.com/forum/#!searchin/rtg-users/vcfstats/rtg-users/-eFsSbWF6ks/1HrnevHTAgAJ):\n",
    "\n",
    ">For the insertions/deletions/indels the table is based on the delta in length rather than total length (which really matters for the indels):\n",
    "\n",
    ">Insertions (pure addition of bases)\n",
    ">A -> AT (length 1 insertion)\n",
    ">ATT -> ATTTT (length 2 insertion) \n",
    "\n",
    ">Deletions (pure deletion of bases)\n",
    ">AT -> A (length 1 deletion)\n",
    ">ATTTT -> ATT (length 2 deletion)\n",
    "\n",
    ">Indels (length changing but not pure)\n",
    ">ATT -> CTTT (length 1 indel)\n",
    ">CTTT -> ATT (length 1 indel)\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "indels = bq.Query('SELECT sample, SUM(DISTINCT CAST(value AS INT64)) AS cum_indels ' +\n",
    "                  'FROM `' + mvp_vcfstats_table + '` ' +\n",
    "                  'WHERE dimension = \"insertions\" ' +\n",
    "                  'OR dimension = \"deletions\" ' +\n",
    "                  'OR dimension = \"indels\" ' +\n",
    "                  'GROUP BY sample').execute().result().to_dataframe()\n",
    "indels.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots()\n",
    "hist = indels.hist(alpha=0.5, ax=ax)\n",
    "ax.set_title(\"Distribution of INDELs per Genome\")\n",
    "ax.set_xlabel(\"INDEL Count\")\n",
    "ax.set_ylabel(\"Frequency\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Frequency of Ti/Tv ratios"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dimension = \"ti_tv_ratio\"\n",
    "ti_tv = bq.Query('SELECT sample, dimension, ' + \n",
    "                 'CAST(value AS FLOAT64) as value ' + \n",
    "                 'FROM `' + mvp_vcfstats_table + '` ' + \n",
    "                 'WHERE dimension = \"' + dimension + '\"').execute().result().to_dataframe()\n",
    "ti_tv.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots()\n",
    "hist = ti_tv.hist(alpha=0.5, ax=ax)\n",
    "ax.set_title(\"Distribution of Ti/Tv ratios\")\n",
    "ax.set_xlabel(\"Ti/Tv Ratio\")\n",
    "ax.set_ylabel(\"Frequency\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.figure()\n",
    "plt.boxplot(ti_tv['value'], 0, 'gD')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Frequency of SNP het/hom ratios"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "snp_het_hom_ratio = bq.Query('SELECT sample, dimension, ' + \n",
    "                             'CAST(value AS FLOAT64) as value ' + \n",
    "                             'FROM `' + mvp_vcfstats_table + '` ' + \n",
    "                             'WHERE dimension = \"snp_het_hom_ratio\"').execute().result().to_dataframe()\n",
    "snp_het_hom_ratio.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots()\n",
    "hist = snp_het_hom_ratio.hist(alpha=0.5, ax=ax)\n",
    "ax.set_title(\"Distribution of SNP het/hom ratios\")\n",
    "ax.set_xlabel(\"Het/Hom Ratio\")\n",
    "ax.set_ylabel(\"Frequency\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# MVP Phase 2 Samtools Flagstat Results (n=1902)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "mvp_flagstat_table = \"\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Percent of reads mapped to reference"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dimension = \"mapped_reads_perc\"\n",
    "mapped_reads_perc = bq.Query('SELECT sample, dimension, ' + \n",
    "                             'CAST(value AS FLOAT64) as value ' + \n",
    "                             'FROM `' + mvp_flagstat_table + '` ' + \n",
    "                             'WHERE dimension = \"' + dimension + '\"').execute().result().to_dataframe()\n",
    "mapped_reads_perc.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots()\n",
    "hist = mapped_reads_perc.hist(alpha=0.5, ax=ax)\n",
    "ax.set_title(\"Distribution of Mapped Reads (%)\")\n",
    "ax.set_xlabel(\"Percent Mapped Reads\")\n",
    "ax.set_ylabel(\"Frequency\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Count of reads mapped to reference"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dimension = \"mapped_reads_count\"\n",
    "mapped_reads_count = bq.Query('SELECT sample, dimension, ' + \n",
    "                              'CAST(value AS FLOAT64) as value ' + \n",
    "                              'FROM `' + mvp_flagstat_table + '` ' + \n",
    "                              'WHERE dimension = \"' + dimension + '\"').execute().result().to_dataframe()\n",
    "mapped_reads_count.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots()\n",
    "hist = mapped_reads_count.hist(alpha=0.5, ax=ax)\n",
    "ax.set_title(\"Distribution of Mapped Reads (Count)\")\n",
    "ax.set_xlabel(\"Count of Mapped Reads (Billions)\")\n",
    "ax.set_ylabel(\"Frequency\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Percent of reads properly paired"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dimension = \"properly_paired_perc\"\n",
    "properly_paired_perc = bq.Query('SELECT sample, dimension, ' + \n",
    "                                'CAST(value AS FLOAT64) as value ' + \n",
    "                                'FROM `' + mvp_flagstat_table + '` ' + \n",
    "                                'WHERE dimension = \"' + dimension + '\"').execute().result().to_dataframe()\n",
    "properly_paired_perc.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots()\n",
    "hist = properly_paired_perc.hist(alpha=0.5, ax=ax)\n",
    "ax.set_title(\"Distribution of Properly Paired Reads (%)\")\n",
    "ax.set_xlabel(\"Percent Properly Paired Reads\")\n",
    "ax.set_ylabel(\"Frequency\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Count of reads properly paired"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dimension = \"properly_paired_count\"\n",
    "properly_paired_count = bq.Query('SELECT sample, dimension, ' + \n",
    "                                 'CAST(value AS FLOAT64) as value ' + \n",
    "                                 'FROM `' + mvp_flagstat_table + '` ' + \n",
    "                                 'WHERE dimension = \"' + dimension + '\"').execute().result().to_dataframe()\n",
    "properly_paired_count.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots()\n",
    "hist = properly_paired_count.hist(alpha=0.5, ax=ax)\n",
    "ax.set_title(\"Distribution of Properly Paired Reads (Count)\")\n",
    "ax.set_xlabel(\"Count of Properly Paired Reads (Billions)\")\n",
    "ax.set_ylabel(\"Frequency\")\n",
    "plt.show()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
